### PR TITLE
Support for custom reporters and build artifacts

### PR DESCRIPTION
**Pull request by nagpalkaran95**
### User facing change log
* Support for passing `reporter` and `reporter-options` arguments
* Support for downloading build artifacts (automatically when running with `--sync`)
* New command`generate-downloads` added for downloading artifacts manually once the build is finished

### CODE REVIEW

File: bin/helpers/buildArtifacts.js [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+    for(let i = 0; i < combs.length; i++) {
+      let comb = combs[i];
+      let sessions = Object.keys(data[comb]);
+      for(let j = 0; j < sessions.length; j++) {
+        let sessionId = sessions[j];
+        let filePath = path.join('./', 'build_artifacts', buildId, comb, sessionId);
+        let fileName = 'build_artifacts.zip';
+        process.env.BUILD_ARTIFACTS_TOTAL_COUNT = Number(process.env.BUILD_ARTIFACTS_TOTAL_COUNT) + 1
+        all_promises.push(downloadAndUnzip(filePath, fileName, data[comb][sessionId]).catch((error) => {
+          process.env.BUILD_ARTIFACTS_FAIL_COUNT = Number(process.env.BUILD_ARTIFACTS_FAIL_COUNT) + 1;
```
Comment (karanshah-browserstack) [inline bin/helpers/buildArtifacts.js:28]: Use a global variable instead of process.env

File: bin/helpers/buildArtifacts.js [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+        messageType = Constants.messageTypes.SUCCESS;
+        message = Constants.userMessages.DOWNLOAD_BUILD_ARTIFACTS_SUCCESS.replace('<build-id>', buildId).replace('<user-path>', process.cwd());
+        logger.info(message);
+      }
+
+      await sendUpdatesToBstack(bsConfig, buildId, args, options);
+      utils.sendUsageReport(bsConfig, args, message, messageType, null);
+  } catch (err) {
+    messageType = Constants.messageTypes.ERROR;
+    errorCode = 'api_failed_build_artifacts';
```
Comment (karanshah-browserstack) [inline bin/helpers/buildArtifacts.js:199]: Fetch this from utils.getErrorCodeFromMsg and fetch it from there.

Comment (nagpalkaran95) [reply bin/helpers/buildArtifacts.js:199]: discussed f2f. not needed

**Review by karanshah-browserstack — APPROVED** [review]

File: bin/commands/generateDownloads.js [hunk 3]
```diff
@@ -0,0 +1,38 @@
+'use strict';
+
+const logger = require("../helpers/logger").winstonLogger,
+      Constants = require("../helpers/constants"),
+      utils = require("../helpers/utils"),
+      downloadBuildArtifacts = require('../helpers/buildArtifacts').downloadBuildArtifacts;
+
+
+module.exports = async function generateDownloads(args) {
```
Comment (karanshah-browserstack) [inline bin/commands/generateDownloads.js:9]: Test cases for this file are missing.

**Review by karanshah-browserstack — CHANGES_REQUESTED** [review]

**Review by karanshah-browserstack — APPROVED** [review]

**Review by karanshah-browserstack — APPROVED** [review]

File: bin/helpers/buildArtifacts.js [hunk 4]
```diff
@@ -28,8 +28,12 @@ const parseAndDownloadArtifacts = async (buildId, data) => {
         let fileName = 'build_artifacts.zip';
         BUILD_ARTIFACTS_TOTAL_COUNT += 1;
         all_promises.push(downloadAndUnzip(filePath, fileName, data[comb][sessionId]).catch((error) => {
-          BUILD_ARTIFACTS_FAIL_COUNT = BUILD_ARTIFACTS_FAIL_COUNT + 1;
-          reject(error);
+          BUILD_ARTIFACTS_FAIL_COUNT += 1;
+          // delete malformed zip if present
+          let tmpFilePath = path.join(filePath, fileName);
```
Comment (karanshah-browserstack) [inline bin/helpers/buildArtifacts.js:33]: Do we need to modify the test cases for this?

**Review by karanshah-browserstack — APPROVED** [review]

File: bin/helpers/buildArtifacts.js (hunk 4 above)
Comment (nagpalkaran95) [reply bin/helpers/buildArtifacts.js:33]: Nope.

**Review by karanshah-browserstack — APPROVED** [review]

**Pull request merged**